sdk/src/main/java/io/customer/sdk/api/HttpRequestRunner.kt

```diff
@@ -0,0 +1,117 @@
+        when (val statusCode = response.code()) {
+            in 500 until 600 -> {
+                val sleepTime = retryPolicy.nextSleepTime
+                return if (sleepTime != null) {
+                    logger.debug("Encountered $statusCode HTTP response. Sleeping $sleepTime seconds and then retrying.")
+
+                    retryPolicyTimer.scheduleAndCancelPreviousSuspend(sleepTime) {
+                        this.performAndProcessRequest(makeRequest)
+                    }
+                } else {
+                    pauseHttpRequests()
+                    prepareForNextRequest() // after retry policy is finished, reset to prepare for the next HTTP request
+
+                    Result.failure(CustomerIOError.ServerDown())
+                }
+            }
+            401 -> {
+                pauseHttpRequests()
+
+                return Result.failure(CustomerIOError.Unauthorized())
+            }
+            else -> {
+                var errorMessage = "No error body from API."
+
+                // by calling .string(), you are not able to get the error body again. retrofit clears the error body after calling .string()
+                response.errorBody()?.string()?.let { errorBodyString ->
+                    errorMessage = errorBodyString // if we can't parse the error body json, the raw response string is good to capture.
+
+                    parseCustomerIOErrorBody(errorBodyString)?.message?.let { parsedErrorMessage ->
+                        errorMessage = parsedErrorMessage
+                    }
+                }
+
+                val customerIOError = CustomerIOError.UnsuccessfulStatusCode(statusCode, errorMessage)
+
+                logger.error("4xx HTTP status code response. Probably a bug? $errorMessage")
+
+                return Result.failure(customerIOError)
+            }
+        }
+    }
+
+    internal fun parseCustomerIOErrorBody(errorBody: String): Throwable? {
+        return jsonAdapter.fromJsonOrNull<CustomerIOApiErrorResponse>(errorBody)?.throwable
+            ?: jsonAdapter.fromJsonOrNull<CustomerIOApiErrorsResponse>(errorBody)?.throwable
+    }
+
+    private fun prepareForNextRequest() {
+        retryPolicy.reset() // in case retry policy was used, reset it so the next request can use it.
+        // do not edit the HTTP pausing. Let that get modified somewhere else and get reset by timing out.
+    }
+
+    // In certain scenarios, it makes sense for us to pause making any HTTP requests to the
+    // Customer.io API. Because HTTP requests are performed by the background queue, there is
+    // a chance that the background queue could make a lot of HTTP requests in
+    // a short amount of time from lots of devices. This makes a performance impact on our API.
+    // By pausing HTTP requests, we mitigate the chance of customer devices causing harm to our API.
+    internal fun pauseHttpRequests() {
```

Comment:
I am having a bit of a hard time understanding why would we need to `pause` making any HTTP requests?

Shouldn't be either rate-limiting from the start so this issue never arise? if we throttle our requests when making from BQ, we wont get to the point that it starts impacting our APIs performance? Thoughts? 

Maybe we can take this to the squad as well for their input on this. ?

Reply:
This has been discussed outside of GitHub. We decided to move forward with this implementation but create a ticket to see if we can improve it in the future.